package tree;

import java.util.ArrayList;
import java.util.List;

import data.Attribute;
import data.Data;

abstract class SplitNode extends Node implements Comparable<SplitNode> {

    class SplitInfo {
        Object splitValue;
        int beginIndex;
        int endIndex;
        int numberChild;
        String comparator = "=";

        SplitInfo(Object splitValue, int beginIndex, int endIndex, int numberChild) {
            this.splitValue = splitValue;
            this.beginIndex = beginIndex;
            this.endIndex = endIndex;
            this.numberChild = numberChild;
        }

        SplitInfo(Object splitValue, int beginIndex, int endIndex, int numberChild, String comparator) {
            this.splitValue = splitValue;
            this.beginIndex = beginIndex;
            this.endIndex = endIndex;
            this.numberChild = numberChild;
            this.comparator = comparator;
        }

        int getBeginIndex() {
            return beginIndex;
        }

        int getEndIndex() {
            return endIndex;
        }

        Object getSplitValue() {
            return splitValue;
        }

        @Override
        public String toString() {
            return "child " + numberChild + " split value" + comparator + splitValue + "[Examples:" + beginIndex + "-" + endIndex + "]";
        }

        String getComparator() {
            return comparator;
        }
    }

    private Attribute attribute;
    protected List<SplitInfo> mapSplit = new ArrayList<SplitInfo>();
    private double splitVariance;

    abstract void setSplitInfo(Data trainingSet, int beginExampleIndex, int endExampleIndex, Attribute attribute);

    abstract int testCondition(Object value);

    SplitNode(Data trainingSet, int beginExampleIndex, int endExampleIndex, Attribute attribute) {
        super(trainingSet, beginExampleIndex, endExampleIndex);
        this.attribute = attribute;
        trainingSet.sort(attribute, beginExampleIndex, endExampleIndex);
        setSplitInfo(trainingSet, beginExampleIndex, endExampleIndex, attribute);

        splitVariance = 0;
        for (SplitInfo splitInfo : mapSplit) {
            double localVariance = new LeafNode(trainingSet, splitInfo.getBeginIndex(), splitInfo.getEndIndex()).getVariance();
            splitVariance += localVariance;
        }
    }

    Attribute getAttribute() {
        return attribute;
    }

    @Override
    double getVariance() {
        return splitVariance;
    }

    @Override
    int getNumberOfChildren() {
        return mapSplit.size();
    }

    SplitInfo getSplitInfo(int child) {
        return mapSplit.get(child);
    }

    String formulateQuery() {
        String query = "";
        for (int i = 0; i < mapSplit.size(); i++) {
            query += i + ":" + attribute + mapSplit.get(i).getComparator() + mapSplit.get(i).getSplitValue() + "\n";
        }
        return query;
    }

    @Override
    public int compareTo(SplitNode o) {
        if (splitVariance < o.splitVariance)
            return -1;
        if (splitVariance > o.splitVariance)
            return 1;
        return 0;
    }

    @Override
    public String toString() {
        String v = "SPLIT : attribute=" + attribute + " " + super.toString() + " Split Variance: " + getVariance() + "\n";
        for (SplitInfo splitInfo : mapSplit) {
            v += "\t" + splitInfo + "\n";
        }
        return v;
    }
}
